Category Widget

jesusbotella opened the pull request:
This PR contains Category Widget as a Web Component. The functionality is described in #90.
Styles and colors were taken from Sketch design.

### Example
![Category Widget](https://user-images.githubusercontent.com/4319728/43456226-ec9a6d74-94c2-11e8-8cda-6e1cb2be5d51.gif)

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.scss:
```diff
@@ -0,0 +1,81 @@
+  @include padding(4);
+
+  :root {
+    --description--color: $color-type-20;
```
why :root, why not declaring the css-vars under the `as-category-widget` namespace

jesusbotella replied:
Yes, it would be better doing it that way.

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.scss:
```diff
@@ -0,0 +1,81 @@
+    --bar--height: $bar--height;
+  }
+
+  .as-category-widget__description,
```
we use `&__description` in the styles, I think we should pick one and be consistent all over the project

jesusbotella replied:
Yes, I think so but the thing here is that it is not `.as-category-widget` but `as-category-widget`. So whether we wrap these styles in `.as-category-widget` or we set that class to the host element and transform `as-category-widget` to `.as-category-widget`.

What do you prefer?

IagoLast replied:
oooouch, you are right... so its fine this way

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+  }
+}
+
+export interface Category {
```
should we move this interfaces to its own .ts file?

jesusbotella replied:
I don't know, maybe we can move them to separated files if we use them anywhere else. What do you think?

IagoLast replied:
As a former java user Im used to have classes/Interfaces splitted in different files but let @rubenmoya @ivanmalagon give their opinion

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.scss:
```diff
@@ -0,0 +1,81 @@
+as-category-widget {
+  display: block;
+
+  @include padding(4);
```
do we want to always have padding or do we want the container element to define this padding?

imagine we are placing this inside a `floating-panel` who should be the responsible of this padding? the category widget or the floating panel should define a padding for its content?

jesusbotella replied:
Yes, that's a matter to discuss, because if you want to use it as a standalone component the padding would be needed, right? 🤷‍♂️

IagoLast replied:
right! IMO the component itself should not define any padding/margin...etc IMO that's responsibility of the container/user

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.spec.ts:
```diff
@@ -0,0 +1,185 @@
+      categoryWidget.clearSelection();
+
+      expect(categoryWidget.getSelectedCategories()).toEqual([]);
+      expect(categoryWidget._onCategoriesChanged).toHaveBeenCalled();
```
I think we want to test that the component fires some kind of event when the categories are cleared. I think with this expect, we are [testing the implementation](https://teamgaslight.com/blog/testing-behavior-vs-testing-implementation) but not the desired behavior.

I think we could create 2 new tests for .clearSelection() checking the component should trigger an event with previously selected categories and...with no selected categories should the component trigger an event? 🤔

jesusbotella replied:
It should always trigger an event when categories are updated, to keep the user updated about the state of the widget.

I only need to modify the current test to listen to the fired event.

As I am testing a similar thing in `._onCategoriesChanged` describe section, I am going to rename that section to `Behaviour` and place that test there.

jesusbotella replied:
No, that's not expected, I thought you meant other thing 🤦‍♂️

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    it('should select category if it is not selected', () => {
+      const category = { name: 'Category 1', value: 10, color: '#000' };
+      categoryWidget._toggleCategory(category);
+      expect(categoryWidget._onCategoriesChanged).toHaveBeenCalled();
```
similar issue here, I think testing the real event would result in a more strong test, this way if we change the implementation (ie, renaming _onCategoriesChanged to _onCategoriesChange) the test will fail

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.spec.ts:
```diff
@@ -0,0 +1,185 @@
+      element.categories = exampleCategories;
+      await testWindow.flush();
+
+      const spy = jest.fn();
```
I think if we rename this to categoriesSelectedSpy the test is easier to read

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.spec.ts:
```diff
@@ -0,0 +1,185 @@
+  describe('._getVisibleCategoriesMaximumValue', () => {
+    it('should return maximum value of visible categories', () => {
+      categoryWidget.categories = exampleCategories;
+      categoryWidget.useTotalPercentage = false;
```
what is `useTotalPercentage` doing?

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.html:
```diff
@@ -0,0 +1,33 @@
+      description="Description"></as-category-widget>
+
+    <script>
+      var categoryWidget = document.getElementsByTagName('as-category-widget')[0];
```
```html
var categoryWidget = document.querySelector('as-category-widget');
```

also do the job

jesusbotella replied:
👍

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  private parseCategories() {
+    const newCategories = [...this.categories];
```
Why not using [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)

If stencill dont do it, we can load a polyfill for IE  I think having a clean code it worth it ;D 

```ts
private parseCategories() {
    const categories = [...this.categories];
    const otherCategory = this.categories.find((category: Category) => category.name === OTHER_CATEGORY_NAME);
    if (otherCategory) {
      return { categories, otherCategory };
    }
    return { categories };
  }
```

jesusbotella replied:
I thought about it first but I don't know why I discarded it 🤔. Definitely seems cleaner.

jesusbotella replied:
I have just realized that we can't do that, because I want `otherCategory` to be out of the array.

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  private parseCategories() {
+    const newCategories = [...this.categories];
```
do we need the spread operator here?

jesusbotella replied:
Yes, we need it because we're cloning the array, `[this.categories]` wouldn't work.

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+    );
+  }
+
+  private _getCategoriesTotalValue(categories: object[]) {
```
Isn't categories a Category array?

```ts
private _getCategoriesTotalValue(categories: Category[]) {}
```

jesusbotella replied:
Yes, it is but as you can see `categories` in the class definition is `object[]`, and needs to be like that for props definition. So I think it's easier to be like like that, do you agree?

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+
+  private _getCategoriesTotalValue(categories: object[]) {
+    return categories.reduce(
+      (maximum, currentCategory: Category) => currentCategory.value + maximum, 0
```
This is not the maximun, is the `accumulator` or the `acum`... copy&paste?  👀

IagoLast replied:
we could also specify type `reduce(acum: number, currentCategory: Category) ...`

jesusbotella replied:
I don't like accumulator at all, and even less being shortened. I am gonna put `currentMaximum`.

We don't need to be explicit here, TypeScript infers parameter's type.

jesusbotella replied:
oh sh*t, yeah, I know what you mean 😂. What do you think about `sum`?

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  private _getVisibleCategoriesMaximumValue() {
+    const categories = this.categories.slice(0, this.numberOfVisibleCategories);
```
we could create a function 

```ts
private _getVisibleCategories() {
  return this.categories.slice(0, this.numberOfVisibleCategories);
}
```

Then we could use this funcion here 

```ts
private _getVisibleCategoriesMaximumValue() {
    const visibleCategories = this._getVisibleCategories();

    return visibleCategories.reduce((maximum, currentCategory: Category) => {
      return currentCategory.value > maximum ? currentCategory.value : maximum;
    }, 0);
  }
```

or even

```ts
private _getVisibleCategoriesMaximumValue() {
    return this._getVisibleCategories().reduce((maximum, currentCategory: Category) => {
      return currentCategory.value > maximum ? currentCategory.value : maximum;
    }, 0);
  }
```

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+  @Prop() public showHeader: boolean = true;
+  @Prop() public useTotalPercentage: boolean = false;
+
+  @Event() public categoriesSelected: EventEmitter<string[]>;
```
we could give a prefix/subfix to eventEmitters in components to easily detect them

`categoriesSelectedEmitter`?

jesusbotella replied:
👍

jesusbotella replied:
We can't name it like that because that variable name is used for the name of the event, so it would be `addEventListener('categoriesSelectedEmitter')`

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  private _renderOtherCategory(category: Category, options: CategoryOptions) {
+    const categoryData = category || {
```
since typescript supports default parameters we could use them, but the syntax is not very familiar in this case:

```ts
private _renderOtherCategory(
    category: Category = {
      name: 'Other',
      value: this._getCategoriesTotalValue(
        this.categories.slice(this.numberOfVisibleCategories, this.categories.length)
      )
    },
    options: CategoryOptions) {

    return this._renderCategory(category, { maximumValue: options.maximumValue, isOther: true });
  }
```

jesusbotella replied:
No, I think it's easier to follow as it is right now. That syntax looks weird.

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+  public render() {
+    return [
+      this._renderHeader(),
+      <ul class='as-category-widget__list'>{this._renderCategories()}</ul>,
```
Is this too much? I'd love to have 3 equal methods 
```jsx
private function _renderContent() {
  return <ul class='as-category-widget__list'>{this._renderCategories()}</ul>
}
```

jesusbotella replied:
No, I thought about it too, would be good :)

I feel like I did a poor function ordering in the class, I think it is convoluted somehow 🤷‍♂️

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+      ? this._getCategoriesTotalValue(this.categories)
+      : this._getVisibleCategoriesMaximumValue();
+
+    if (otherCategory || moreCategoriesThanVisible) {
```
wow the logic is not very straightforward here, I wasn't expecting `_renderOtherCategory` to render  a default `others` when `otherCategory` parameter is not given... not sure about the best solution

jesusbotella replied:
I dunno either 🤷‍♂️

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+    const selectedCount = this.selectedCategories.length ? this.selectedCategories.length : 'All';
+
+    return (
+      <footer class='as-category-widget__footer'>
```
not sure if footer is the right tag for this, is not an obligation but a recommendation, see [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/footer), [this](https://www.w3.org/wiki/Html/Elements/footer) or [this](https://www.w3schools.com/tags/tag_footer.asp) 

> The footer tag defines a footer for a document or section. A footer element should contain information about its containing element. typically contains: authorship information, copyright information, contact information, sitemap. back to top links, related documents

jesusbotella replied:
I think it contains information about its containing element, right?
It has the information of how many categories are selected

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  private _isSelected(categoryName: string) {
+    return this.selectedCategories.indexOf(categoryName) > -1;
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

?

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  private _toggleCategory(category: Category) {
+    const newSelectedCategories = this._isSelected(category.name)
```
can we assign directly the result of the ternary operator to this.selectedCategories?

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+    return color;
+  }
+
+  private parseCategories() {
```
_parseCategories?

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  private _renderFooter() {
+    const selectedCount = this.selectedCategories.length ? this.selectedCategories.length : 'All';
```
`const selectedCount = this.selectedCategories.length || 'All'`

IagoLast commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+  private _renderCategory(category: Category, options: CategoryOptions) {
+    const isSelected = this._isSelected(category.name);
+    const isAnyCategorySelected = this.selectedCategories.length > 0;
+    const barColor = !options.isOther
```
why comparing against the negated condition? 😂  and not just:
```js
const barColor = options.isOther
      ? OTHER_CATEGORY_COLOR
      : this._getBarColor(category.color || this.defaultBarColor, { isSelected });
```

jesusbotella replied:
Because I found it more intuitive to put the `_getBarColor` in the upper branch but I don't mind :)

VictorVelarde commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,203 @@
+  @Event() public categoriesSelected: EventEmitter<string[]>;
+
+  @State() private selectedCategories: string[] = [];
+  @State() private numberOfVisibleCategories: number = 5;
```
IMO this should be easily set from outside by the client

jesusbotella replied:
I added a property named `visibleCategories` to do that :) Thank you!

jesusbotella commented:
I think I have made a bad decision when it comes to showing categories, I should have allowed to show all categories by default. I am going to refactor the logic to allow that.

IagoLast commented on packages/components/src/components/as-category-widget/README.md:
```diff
@@ -0,0 +1,190 @@
+---
+<as-category-widget
+  heading="Business Volume"
+  description="Description"></as-category-widget>
```
we could pass all possible parameters here with the default values

IagoLast commented on packages/components/src/components/as-category-widget/README.md:
```diff
@@ -0,0 +1,190 @@
+
+### Props
+
+#### **categories** (array required)
```
I think the most common way to document APIS in markdown is:

Required parameter:
name:type

Optional parameter:
[name:type]=default


cc @ivanmalagon @rubenmoya

piensaenpixel commented on packages/components/src/components/as-category-widget/as-category-widget.scss:
```diff
@@ -0,0 +1,81 @@
+    background-color: $color-ui-20;
+  }
+
+  .as-category-widget__value {
```
What do you think if we avoid nesting the classes? i know it's more organized ... but I'd like a clearer css output

jesusbotella replied:
Nesting inside `as-category-widget`? I don't really mind putting them outside `as-category-widget`. We're using BEM so there shouldn't be any collisions :)

What do you think @IagoLast @rubenmoya @ivanmalagon?

piensaenpixel commented on packages/components/src/components/as-category-widget/as-category-widget.scss:
```diff
@@ -0,0 +1,81 @@
+    position: relative;
+
+    &::after {
+      content: attr(data-value);
```
I'm not sure about using the pseudoelement for showing content in the html

jesusbotella replied:
Yes, I didn't think about accessibility 😞.
And leveraging this situation I think I am going to make an improvement and use an ellipsis to cut overflowed text :)

jesusbotella commented:
I changed @piensaenpixel suggestions and added an ellipsis to category title, so that when there's more text than maximum width there are `...` cutting it.

<img width="397" alt="screen shot 2018-08-01 at 13 10 47" src="https://user-images.githubusercontent.com/4319728/43518959-fc9bbd38-958d-11e8-8bf7-f40d76f32e1f.png">

rubenmoya commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,210 @@
+    const { categories, otherCategory } = this._parseCategories();
+    let otherCategoryTemplate;
+
+    const categoriesToRender =  moreCategoriesThanVisible
```
I think we can avoid this ternary you can always do `categories.slice(0, this.visibleCategories)`, if the array is smaller it will return it with the same size, in fact I think we can just use `this._getVisibleCategories`

rubenmoya commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,210 @@
+    return (
+      <footer class='as-category-widget__footer'>
+        <div class='as-category-widget__count as-body'>{selectedCount} selected</div>
+        { this.showClearButton
```
What I use to do in react was:

```
{ this.showClearButton && <button class='as-category-widget__clear' onClick={() => this.clearSelection()}>Clear selection</button> }
```

I think it looks cleaner, but it's just my opinion 😂

rubenmoya commented on packages/components/src/components/as-category-widget/as-category-widget.tsx:
```diff
@@ -0,0 +1,210 @@
+  private _renderCategory(category: Category, options: CategoryOptions) {
+    const isSelected = this._isSelected(category.name);
+    const isAnyCategorySelected = this.selectedCategories.length > 0;
+    const barColor = options.isOther
```
Can't we move the `isOther` logic to the `getBarColor` method? In my head the `renderCategory`method shouldn't do this work, I expect `getBarColor` to be smart enough to give me the final color, also, we are doing something similar with the `isSelected` what do you think?

jesusbotella replied:
Yes, it would be so much better :)

rubenmoya commented on packages/components/src/utils/styles.js:
```diff
@@ -0,0 +1,25 @@
+// pSBC.js
+// From Pimped JavaScript (https://github.com/PimpTrizkit/PJs)
+// Usage: https://github.com/PimpTrizkit/PJs/wiki/12.-Shade,-Blend-and-Convert-a-Web-Color-(pSBC.js)
+export function shadeOrBlend(p, from, to) {
```
<a href="https://imgflip.com/i/2f59e6"><img src="https://i.imgflip.com/2f59e6.jpg" title="made at imgflip.com"/></a>

ivanmalagon commented on packages/components/src/components/as-category-widget/as-category-widget.spec.ts:
```diff
@@ -0,0 +1,202 @@
+      element.showClearButton = true;
+      await testWindow.flush();
+
+      expect(element).toMatchSnapshot();
```
What is the snapshot that is compared to? Where?

IagoLast replied:
https://jestjs.io/docs/en/snapshot-testing

jesusbotella replied:
That's basically it. It creates the references in a folder, which is commited to the repo, to be able to compare them in every test run. If there's something different, it fails.

jesusbotella commented:
I have just added JSDocs that I completely forgot about 😂, and added two more examples at the bottom. One using total sum percentage calculation, and other showing only category bars.
